content/extending-chat-widget/custom-identity-provider/index.mdx

```diff
@@ -0,0 +1,174 @@
+
+```js
+    // Custom Identity Provider
+    window.__lc.custom_identity_provider = function() {
+      return {
+          getToken: ...,
+          getFreshToken: ...,
+          hasToken: ...,
+          invalidate: ...
+      }
+    }
+```
+
+# Methods
+
+The aforementioned function assigned to `custom_identity_provider` property should return a set of the following promises:
+
+- `getToken` - resolving [Chat Widget token](#chat-widget-token). If you want to cache the token somehow, this should return the cached token instead of a fresh request to API.
+- `getFreshToken` - resolving [Chat Widget token](#chat-widget-token). This should always make a call for a fresh token from accounts API.
+- `hasToken` - resolving boolean. Determining whether a token has been acquired.
+- `invalidate` - resolving nothing. When called should remove the current token. There is no need to do anything else, as a new token will be requested by `getFreshToken` afterward.
+
+# Chat Widget Token
+
+LiveChat authorization token you are willing to resolve in functions should contain the following properties:
+
+| Parameter      | Description                                                                      |
+| -------------- | -------------------------------------------------------------------------------- |
+| `accessToken`  | A token you can use to call LiveChat APIs on behalf of the user (customer).      |
+| `entityId`     | The ID of the customer.                                                              |
+| `expiresIn`    | A number in **milliseconds** specifying how long the `accessToken` will be valid.|
+| `tokenType`    | Value: `Bearer`.                                                                 |
+| `creationDate` | Unix timestamp specyfing the time of creation of the token.                       |
+| `licenseId`    | LiveChat license ID.                                                             |
+
+## How to acquire it?
+
+In order to get the token you need to implement authorizing function making a call to a `/customer` endpoint from LiveChat Accounts API. To do so, you can follow [creating a new customer flow](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#creating-a-new-customer) or have a look at our [example](https://github.com/livechat/identity-provider-example/blob/master/lib/get-customer-token.ts) using cookies. Mind the fact that the [data](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#response-4) returned from this endpoint is not 1:1 with the token shape described above. You will have to add the `creationDate` and `licenseId` fields. Additionally, we expect `expiresIn` to be defined in milliseconds while the endpoint operates in seconds. Therefore you should multiply the provided value by `1000` before passing it further.
```

Comment:
```suggestion
In order to get the token, you need to implement a function that follows authorization flow making a call to a `/customer` endpoint from LiveChat Accounts API. To do so, you can follow [creating a new customer flow](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#creating-a-new-customer) or have a look at our [example](https://github.com/livechat/identity-provider-example/blob/master/lib/get-customer-token.ts) using cookies. Mind the fact that the [data](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#response-4) returned from this endpoint is not 1:1 with the token shape described above. You will have to add the `creationDate` and `licenseId` fields. Additionally, we expect `expiresIn` to be defined in milliseconds while the endpoint operates in seconds. Therefore you should multiply the provided value by `1000` before passing it further.
```

Reply:
We should mention that such "function" needs to be a backend service

Reply:
@konradk It is not true, it does not need to be called server-side. The customer authorization flow based on cookies does not require excachnging the `code` with corresponding `client_secret` for the token as it is done in the flow based on agent credentials. It is 100% doable on the client-side and we are doing exactly this in the CW.

The flow is as simple as: call `POST: http://api.livechat.com/v3.3/customer/token` with mainly `client_id` and `license` in the paylod and grab the token from the response. Cookies are being automatically set from the response by the browser so they will be also sent together with the next request to refresh the already existing customer identity.

The reason why we are doing it server-side in the `identity-provider-example` app is that we want to bind both identities: the service User and LiveChat Customer, together and provide Chat Widget with the appropriate Customer token for currently signed in User.

Reply:
In other words, you can use it totally on client-side and for example, use customer API to extend the capabilities of stock CW without the need for dedicated backend service. It is actually a pretty great use case for this feature that it allows extending the CW without the need to write it from scratch using `cusotmer-sdk` to handle some simple extended cases.